pkg/arvo/neo/cod/std/src/fil/messenger.md

```diff
@@ -0,0 +1,649 @@
+    :_  :-  %groupchat
+        !>(sta(pending (~(put in pending.sta) ship.poke)))
+    :~  :-  provider.poke
+        [%poke groupchat-diff/!>([%invited here.bowl])]
+    ==
+  ::
+      ::  remove someone from chat. this only removes their ability to post;
+      ::  they'll still be receiving new messages!
+      %remove
+    ?>  =(our ship.src):bowl
+    ?>  (~(has in members.sta) ship.poke)
+    :-  ~
+    :-  %groupchat
+    !>  %=  sta
+          pending  (~(del in pending.sta) ship.src.bowl)
+          members  (~(del in members.sta) ship.src.bowl)
+        ==
+  ::
+      :: when invitee acks, remove them from pending
+      :: and add them to pub's permissions
+      %acked
+    ?>  (~(has in pending.sta) ship.src.bowl)
+    :-  ~
+    :-  %groupchat
+    !>  %=  sta
+          pending  (~(del in pending.sta) ship.src.bowl)
+          members  (~(put in members.sta) ship.src.bowl)
+        ==
+  ::
+      %post-to-host
+    :_  state
+    :~  :-  host.sta
+        [%poke groupchat-diff/!>([%host-to-pub text.poke])]
+    ==
+  ::
+      %host-to-pub
+    ?>  (~(has in members.sta) ship.src.bowl)
+    :_  state
+    :~  :-  (snoc here.bowl %pub)
+        [%poke message/!>([ship.src.bowl now.bowl text.poke])]
+    ==
+  ==
+```
+
+### Messenger shrub
+`/imp/messenger` is the top-level interface through which users can create, post in, and manage groupchats and DMs. This is the only part of the system that has a UI.
+
+#### kook:neo
+Messenger has no state. This shrub is just an interface for creating groupchats and DMs, which are its kids. If those kids are `/imp/dm`s they take `%dm-diff`s, and if they're `/imp/groupchat`s they take `%groupchat-diff`s.
+
+{what's going on with the `%t` piths?}
+
+```hoon
+++  state  [%pro %sig]
+++  poke  (sy %dm-diff %groupchat-diff ~)
+++  kids
+  %-  some
+  :-  %y
+  %-  ~(gas by *lads:neo)
+  :~  :-  [[%.n %t] %.n]
+      [[%pro %dm] (sy %dm-diff ~)]
+      :-  [[%.n %t] %.n]
+      [[%pro %groupchat] (sy %groupchat-diff ~)]
+  ==
+++  deps  *deps:neo
+++  form
+  ...
+```
+
+Messenger only supports three user actions: creating a new DM, creating a new groupchat, and inviting someone to a groupchat. Everything else is handled by this shrub's kids.
+
+```hoon
+$%  [%new-dm partner=ship]
+    [%new-groupchat name=@t]
+    [%invite-to-groupchat name=@t =ship]
+==
+```
+
+#### +init
+Messenger does nothing much on `+init`. Here's `%sig` again, this time head-tagging an empty `pail:neo`. This shrub stores no state, so we can't inject some when we initialize it.
+
+```hoon
+++  init
+  |=  old=(unit pail:neo)
+  ^-  (quip card:neo pail:neo)
+  [~ [%sig !>(~)]]
+```
+
+#### +poke
+`/imp/messenger` takes `%messenger-diff`s, but it also takes `%dm-diff`s and `%groupchat-diff`s and, if they're invites to those chats, `%make`s the chat at the right location in the namespace.
+
+{what is the difference between `%invite-to-groupchat` and `%invited` `%groupchat-diff`?}
```

Comment:
because only imp/messenger knows who the provider is, because groupchat should be agnostic about where it's installed/used